alf/networks/encoding_networks.py

```diff
@@ -230,6 +230,149 @@ def __init__(self,
+            "channel number mis-match")
+
+        # compute conv shape and padding shape
+        conv_shapes = []
+        out_paddings = []
+        out_shape = output_shape[1:]
+        for i, paras in enumerate(transconv_layer_params[::-1]):
+            filters, kernel_size, stride = paras[:3]
+            kernel_size = common.tuplify2d(kernel_size)
+
+            padding = paras[3] if len(paras) > 3 else 0
+            padding = common.tuplify2d(padding)
+            conv_shape = self._conv_out_shape(out_shape, padding, kernel_size,
+                                              stride)
+            out_padding = self._output_padding_shape(
+                out_shape, conv_shape, padding, kernel_size, stride)
+            out_shape = conv_shape
+            conv_shapes.append(conv_shape)
+            out_paddings.append(out_padding)
+
+        input_tensor_spec = TensorSpec((input_size, ))
+
+        assert isinstance(transconv_layer_params, tuple)
+        assert len(transconv_layer_params) > 0
+
+        nets = []
+        if preprocess_fc_layer_params is not None:
+            for size in preprocess_fc_layer_params:
+                nets.append(
+                    layers.FC(
+                        input_size,
+                        size,
+                        activation=activation,
+                        kernel_initializer=kernel_initializer))
+                input_size = size
+
+        start_decoding_shape = [
+            start_decoding_channels, conv_shapes[-1][0], conv_shapes[-1][1]
+        ]
+        nets.append(
+            layers.FC(
+                input_size,
+                np.prod(start_decoding_shape),
+                activation=activation,
+                kernel_initializer=kernel_initializer))
+
+        nets.append(alf.layers.Reshape(start_decoding_shape))
+
+        in_channels = start_decoding_channels
+
+        for i, paras in enumerate(transconv_layer_params):
+
+            filters, kernel_size, strides = paras[:3]
+            padding = paras[3] if len(paras) > 3 else 0
+            output_padding = out_paddings[-(i + 1)]
+
+            act = activation
+            if i == len(transconv_layer_params) - 1:
+                act = output_activation
+
+            nets.append(
+                layers.ConvTranspose2D(
+                    in_channels,
+                    filters,
+                    kernel_size,
+                    activation=act,
+                    kernel_initializer=kernel_initializer,
+                    strides=strides,
+                    padding=padding,
+                    output_padding=output_padding))
+            in_channels = filters
+
+        super().__init__(nets, input_tensor_spec=input_tensor_spec, name=name)
+
+    def _conv_out(self, h_in, padding, kernel_size, stride):
+        return int((h_in + 2. * padding - (kernel_size - 1.) - 1.) / stride +
+                   1.)
+
+    def _output_padding(self, h_in, conv_out, padding, kernel_size, stride):
```

Comment:
Suggest just merging this function into _conv_padding_shape to avoid creating more functions. Same for _conv_out.

Reply:
Merged.